### ISSUE TITLE
cups-browsed: Inappropriate PPD file generated

### ISSUE THREAD

**Issue by debiantriage** [OpenPrinting/cups-filters#152]
I am using cups-browsed 1.25.4-2 on Debian. The generated PPD for an ENVY4500 has the line

> *cupsFilter2: "application/vnd.cups-pdf application/pdf 0 -"

The TXT record has

> pdl=application/vnd.hp-PCL,image/jpeg,application/PCLm,image/urf

Consequently, printing to the device will fail.

This is Debian Bug #939476:

[https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=939476](url)

Regards,

Brian.

**Comment by tillkamppeter** [comment]
Does your print queue point directly to a printer (device URI is `ipp://...` or `ipps://...` or to a remote CUPS queue for the printer (device URI is `implicitclass://...`)?

**Comment by debiantriage** [comment]
`lpstat -t` gives me

> implicitclass://ENVY4500/

Is that what you want?

-- 
Brian.

**Comment by tillkamppeter** [comment]
Yes, this means that the queue with the PDF line added to the PPD file is not pointing to the printer directly, but to a CUPS server which hosts this printer. This means that your queue has to send PDF to the remote CUPS server as that one has the driver for this printer set up. That server converts the PDF into one of the printer's PDLs (PCL or Apple Raster, depends on the PPD which the queue on the remote server uses). So this PPD is correct.
If the printer does not print, we need debug log output of cups-browsed on your client and of CUPS on your remote server.

**Comment by debiantriage** [comment]
There is a single computer on this network with a CUPS server and sharing its queues. I put _BrowseLocalProtocols none_ in its _cupsd.conf_ and restarted cupsd.  _avahi-browse -art_ shows the ENVY4500 as the only printing device on the network. The entry disappeared when the device was switched off.

I have a Debian unstable installation on another partition. It has not been updated and has cups-browsed 1.21.6-3. _lpstat -t_ indeed shows the device UR as _ipp://..._. However, I am convinced that the printer is being directly contacted on both installations.

I have attached debug log outputs for cups-browsed 1.21.6-3 and 1.25.4-2 and note that both Remote printer overview sections have an _ipp://..._ URI.

An everywhere queue is correctly set up without a *cupsFilter2 line for PDF (as is a non-auto driverless queue).

Regards,

Brian.
[debug_log-1.21.6-3.txt](https://github.com/OpenPrinting/cups-filters/files/3583971/debug_log-1.21.6-3.txt)
[debug_log-1.25.4-2.txt](https://github.com/OpenPrinting/cups-filters/files/3583973/debug_log-1.25.4-2.txt)

**Comment by tillkamppeter** [comment]
The print queue which cups-browsed 1.25.4 has generated is correct. It points directly to the printer (the URI you see in the overview in the debug log of cups-browsed), The URIs of the local queues are now always "implicitclass://...". This is to allow also to cluster native IPP printers and not only remote CUPS queues. Also the cupsFilter line for PDF in the PPD is correct as for clustering support of printers with different PDLs the cups-browsed-generated queues are supposed to always send only PDF to the backend (which is the implicitclass backend). The implicitclass backend then completes the filtering process to output the PDL which the printer needs.
So nothing wrong up to now.
If the printer does not print, please provide a CUPS error_log (in debug mode) of your local CUPS.
@deepak0405, can you also have a look here?

**Comment by debiantriage** [comment]
brian@test:~$ lpstat -t
scheduler is running
no system default destination
device for ENVY4500: implicitclass://ENVY4500/
ENVY4500 accepting requests since Fri 06 Sep 2019 14:50:26 BST
printer ENVY4500 is idle.  enabled since Fri 06 Sep 2019 14:50:26 BST
brian@test:~$
brian@test:~$ lp -d ENVY4500 hello
request id is ENVY4500-2 (1 file(s))
brian@test:~$
brian@test:~$ lpstat -t
scheduler is running
no system default destination
device for ENVY4500: implicitclass://ENVY4500/
ENVY4500 accepting requests since Fri 06 Sep 2019 14:55:56 BST
printer ENVY4500 is idle.  enabled since Fri 06 Sep 2019 14:55:56 BST
        Backend failed
ENVY4500-2              brian             1024   Fri 06 Sep 2019 14:55:44 BST

(I do not know why some lines are formatted struck out).

Regards,

Brian.

[error_log.txt](https://github.com/OpenPrinting/cups-filters/files/3584321/error_log.txt)

**Comment by tillkamppeter** [comment]
@debiantriage, did you install cups-filters from regular Debian packages? Are all packages of cups-filters of the same version?
@deepak0405, there seems to be something wrong with what cups-browsed is passing to the implicitclass backend, could you have a look into the error_log.txt attached to the previpus comment of @debiantriage? Thanks.

**Comment by deepak0405** [comment]
@debiantriage Can you also send the `cups-browsed` error log when a job is sent to the printer.

**Comment by debiantriage** [comment]
@tillkamppeter All the packages on the machine are from the Debian unstable archive. The cups-filters and cups-browsed packages are all version 1.25.4-2.

@deepak0405 cups-browsed error log attached. I tried to print again. The CUPS error_log now says the implicitclass backend stopped with status 1. Previously it said "crashed on status 11".

I am also having to rest the printer each time as it does not seem to like what is sent to it. Please ask if you need more logs.

Regards,

Brian.

[cups-browsed_log-2.txt](https://github.com/OpenPrinting/cups-filters/files/3584740/cups-browsed_log-2.txt)

**Comment by deepak0405** [comment]
@tillkamppeter The value of `dest_host` was not set in implicitclass.c which was causing backend failure rather than sending an error. This is fixed in 1b1d03df21f3ac4db4a0b7b39dcfac0249f7018f. The log shows that the job is not sent the second time due to timeout, but the reason was job was not completed in the first time itself is not clear.

@debiantriage The cups-browsed_log-2 shows that the job request was done twice, the error_log.txt file contains the `log` only for the second time the request was made. Can you please do a request once more and send both the cups-browsed logs and cups log.

**Comment by debiantriage** [comment]
Here you are. The CUPS error_log once again shows the filters generating a PDF, which is not a PDL the printer processes.

Cheers,

Brian.

[error_log-3.txt](https://github.com/OpenPrinting/cups-filters/files/3586587/error_log-3.txt)
[cups-browsed_log-3.txt](https://github.com/OpenPrinting/cups-filters/files/3586588/cups-browsed_log-3.txt)

**Comment by deepak0405** [comment]
The `error_log` shows that the print job contains two files or may be the document is printed in two steps. 
`Sending file 2 using chunking...`

@debiantriage What document you are printing ? If possible, Can you post the document, it will help me reproduce the error.

**Comment by debiantriage** [comment]
"chunking" is a word I associate with processing a PDF. But I am really out of my league here.

The document was produced by the text editor, vim. It has the single word "hello" in it.

Cheers,

Brian.

**Comment by debiantriage** [comment]
I have installed Ubuntu eoan (with cups-browsed 1.25.4-1) from its mini.iso; base system only. Then installed the cups package. The outcome is unchanged.

Cheers,

Brian.

**Comment by deepak0405** [comment]
@debiantriage @tillkamppeter   I tried reproducing the error following these steps : 
1.  Made a `hello` file which contains only one world "hello".
2. Created a IPP Network printer using ippserver.
3. Sent the job to the printer using  `lp printer_name hello`

But the job was completed in my case, did I miss anything ?

@debiantriage  Please  send output of 
                `ipptool -tv <URI> get-printer-attributes-2.0.test`
https://github.com/steveathon/cups/blob/master/test/get-printer-attributes-2.0.test, I will try to simulate  a printer with the same capabilities as `ENVY4500`

cups-filters version : 1.25.5

**Comment by debiantriage** [comment]
Thanks for going to this trouble, deepak0405. The attributes test file I have on my system is the same as the one you pointed to.

Cheers,

Brian.

[envy4500-attributes.txt](https://github.com/OpenPrinting/cups-filters/files/3587975/envy4500-attributes.txt)

**Comment by deepak0405** [comment]
Hi @debiantriage,
Sorry I wrote the wrong command, I need the attribute file in  `ippserver` attribute file format. Please send the output of `ipptool` command using `--ippserver filename` flag.

**Comment by debiantriage** [comment]
@deepak0405 No problem. I used `ipptool --ippserver envy4500-
[envy4500-attributes.txt](https://github.com/OpenPrinting/cups-filters/files/3594400
/envy4500-attributes.txt)`

[envy4500-attributes.txt](https://github.com/OpenPrinting/cups-filters/files/3594419/envy4500-attributes.txt)

Cheers,

Brian.

**Comment by deepak0405** [comment]
@tillkamppeter 
The job is not getting completed because ipp backend thinks that it has to print two files but actually we are printing only one file.

The ipp backend uses `argc-6` to calculate the number of files which are to be printed. In this case, the argc is 8, but we are sending only 7 arguments from `implicitclass`.
```
D [07/Sep/2019:10:12:48 +0100] [Job 9] argv[0]=\"implicitclass://envy4500/\"
D [07/Sep/2019:10:12:48 +0100] [Job 9] argv[1]=\"9\"
D [07/Sep/2019:10:12:48 +0100] [Job 9] argv[2]=\"brian\"
D [07/Sep/2019:10:12:48 +0100] [Job 9] argv[3]=\"hello\"
D [07/Sep/2019:10:12:48 +0100] [Job 9] argv[4]=\"1\"
D [07/Sep/2019:10:12:48 +0100] [Job 9] argv[5]=\"cups-browsed cups-browsed-dest-printer=\\\"7\\ NO_DEST_FOUND\\\" finishings=3 number-up=1 job-uuid=urn:uuid:3cfc82cb-93d1-39ee-7bfa-1bfb5d5e01a1 job-originating-host-name=localhost date-time-at-creation= date-time-at-processing= time-at-creation=1567847567 time-at-processing=1567847567 document-name-supplied=hello output-format=apple-raster Resolution=600dpi media-class=pwg page-logging=on\"
```
The line `cups-browsed-dest-printer=\\\"7\\ NO_DEST_FOUND\\\"` is not supposed to be the part of the job arguments. This option is not present in argv[5] when I print a file from latest cups filters. @tillkamppeter Are we adding cups-browsed-dest-printer in the job attributes somewhere ?

**Comment by tillkamppeter** [comment]
I think you have some quoting problem here. The 5th argument is no properly quoted so that it is considered as 2 arguments. Make also sure that the 5th argument only contains only arguments which the IPP backend or the destination printer actually needs. For example you add the option "cups-browsed-dest-printer" to the cluster queue, so that the implicitclass backend knows about the destination printer URI, the printer's PDL, ... This argument does not need to get passed on to the IPP backend used to send out the job to the actual destination.

**Comment by deepak0405** [comment]
**Root Cause of Error** :
The `scheduler/ipp.c` file sets the default values for the printer. It also sets `cups-browsed-dest-printer` in the default value. When a new job is created this value is fetched and attached to job arguments. Ideally, this should not be the part of job arguments, because `implicitclass` backend fetches this value from the cups-browsed.

@debiantriage I have fixed this issue in https://github.com/deepak0405/cups-filters/tree/solve-implicitclass-error branch. Can you please see that this patch fixes the error ?

**Comment by debiantriage** [comment]
@deepak0405 I purged the Debian cups-browsed package, downloaded the zip file and ran ./autogen, ./configure and make. Then 'make install' , adjusted cups-browsed.conf and started cups-browsed. The outcome is unchanged.

The generated PPD and a log is attached.

Cheers,

Brian.

[ENVY4500.ppd.txt](https://github.com/OpenPrinting/cups-filters/files/3639687/ENVY4500.ppd.txt)
[cups-browsed_log.txt](https://github.com/OpenPrinting/cups-filters/files/3639690/cups-browsed_log.txt)

**Comment by deepak0405** [comment]
@debiantriage  Please also attach the cups error_log file.

**Comment by deepak0405** [comment]
The changes in `cupsFilter` line in ppd file is because of a new feature addition. Now whenever a job is sent to the printer :
1. The `cups` converts the print file to the pdf format. So we have `*cupsFilter2: "application/vnd.cups-pdf application/pdf 0 -"` line in the ppd file by default.

2. Next the job is sent to the `implicitclass` backend, which converts the pdf file to the format supported by the printer and sends the job to the printer.

**Comment by debiantriage** [comment]
>    Can you please try printing a job on this printer ? Whether the job is getting completed ? Also please attach the cups error log for the job.

@deepak0405 Thank you for your patience.

A simple text file printed, so, as far as I am concerned, this report can be closed

Regards,

Brian.

[error_log.txt](https://github.com/OpenPrinting/cups-filters/files/3639837/error_log.txt)

**Comment by debiantriage** [comment]
@deepak0405 Thank you (and Till) for explaining the workings of the implicitclass backend.

>  The changes in cupsFilter line in ppd file is because of a new feature addition.

I have a lot the learn (and unlearn). One of the things in the PPD file that puzzled me was *PCFileName: "ippeve.ppd". I expected "drvless.ppd".

Regards,

Brian.

**Comment by deepak0405** [comment]
@debiantriage Glad to know that the fix worked for you.

@tillkamppeter Ideally the ppd file should contains the information about the filter chain which converts the input file to output supported by the printer. Since we are changing the ideal behaviour, we should document the change somewhere, where it is reachable by a lot of people. Where we can write the same ?

@tillkamppeter 
Also, in `ppdgenerator.c`, we are setting the value of PCFileName as `drvless.ppd` 
```
cupsFilePuts(fp, "*PCFileName: \"drvless.ppd\"\n");
```
But  the `ppd-cache.c` sets the value to `ippeve.ppd`.
```
cupsFilePuts(fp, "*PCFileName: \"ippeve.ppd\"\n");
```
I am not much aware of the flow and the ideal behaviour in this case. Till can you please check ?

**Comment by tillkamppeter** [comment]
@deepak0405, could you post [your fix](https://github.com/deepak0405/cups-filters/commit/ac935005b) as Pull Request?
@deepak0405, any type of documentation should be added to the cups-filters package itself at first, in the currenmt situation to the README file, but on the long run we need to clean up the documentation of cups-filters. Once having done this we should also consider putting important information about cups-filters onto our web site.
@deepak0405, the PCFileName in the PPD files has absolutely no functionality for us, it only needs to be present to fulfill Adobe's PPD specs. It is from the very early times of PostScript, where PCs were still running MS-DOS (Disk Operating System, not Denial Of Service) requiring file names with a max 8 character base name and a max 3 character file name extension. In the PCFileName field Adobe/the PostScript printer manufacturers suggested appropriate file names to printer driver packagers/operating system vendors. So the content of the field does not actually matter.

**Comment by tillkamppeter** [comment]
Fixed via pull request #159.
